test/hotspot/jtreg/serviceability/jvmti/vthread/PopFrameTest/PopFrameTest.java

```diff
@@ -145,6 +145,7 @@ public static void run() {
             }
             log("Main #B.2: got expected JVMTI_ERROR_NONE");
             resumeThread(testTaskThread);
+
```

Comment:
You probably don't need this minor edit or the copyright update any more.

However, it's still unclear to me how the ensureAtBreakpoint() below is suppose to work if we already called notifyAtBreakpoint() between the 1st and 2nd ensureAtBreakpoint(). From what I can tell, notifyAtBreakpoint() clears the flag that ensureAtBreakpoint() checks for, and there is no 2nd breakpoint to set it again. I would expect the ensureAtBreakpoint() below to always wait indefinitely, but that doesn't appear to be the case. So how is this working? I must be missing something.

Reply:
Thanks, Chris. I noticed that some an additional cleanup is needed in the Java part.
The `TestTask.run()` has two calls to the method `B()`:
```
        public void run() {
            log("TestTask.run: started");

            A();
            sleep(1); // to cause yield

            prepareAgent(TestTask.class, false); // No doPopFrame
            B();
            sleep(1); // to cause yield

            prepareAgent(TestTask.class, true); // doPopFrame
            B();
            sleep(1); // to cause yield

            C();
            finished = true;
        }
```
The second `B()` call does the JVMTI `PopFrame` call on its own thread in the `Breakpoint()` event hadler.
Please, note, `true` in the `prepareAgent` parameters:
   `prepareAgent(TestTask.class, true); // doPopFrame`
So, the second `ensureAtBreakpoint()` is a sync point with it.